File: ponio/examples/arenstorf.cpp
// Copyright 2022 PONIO TEAM. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// NOLINTBEGIN(misc-include-cleaner)

#include <cmath>
#include <cstddef>
#include <filesystem>
#include <string>
#include <valarray>

#include <ponio/observer.hpp>
#include <ponio/problem.hpp>
#include <ponio/runge_kutta.hpp>
#include <ponio/solver.hpp>
#include <ponio/splitting.hpp>

// NOLINTEND(misc-include-cleaner)

/*
solve Arenstorf orbit problem
*/

struct arenstorf_model
{
    using state_t = std::valarray<double>;

    double mu;

    arenstorf_model( double m )
        : mu( m )
    {
    }

    void
    phi_1( double, state_t const& y, state_t& dy ) const // NOLINT(readability-convert-member-functions-to-static)
    {
        double const y3 = y[2];
        double const y4 = y[3];

        dy[0] = y3;
        dy[1] = y4;
        dy[2] = 0.;
        dy[3] = 0.;
    }

    void
    phi_2( double, state_t const& y, state_t& dy ) const
    {
        double const y1 = y[0];
        double const y2 = y[1];
        double const y3 = y[2];
        double const y4 = y[3];

        double const r1 = sqrt( ( y1 + mu ) * ( y1 + mu ) + y2 * y2 );
        double const r2 = sqrt( ( y1 - 1 + mu ) * ( y1 - 1 + mu ) + y2 * y2 );

        dy[0] = 0.;
        dy[1] = 0.;
        dy[2] = y1 + 2 * y4 - ( 1 - mu ) * ( y1 + mu ) / ( r1 * r1 * r1 ) - mu * ( y1 - 1 + mu ) / ( r2 * r2 * r2 );
        dy[3] = y2 - 2 * y3 - ( 1 - mu ) * y2 / ( r1 * r1 * r1 ) - mu * y2 / ( r2 * r2 * r2 );
    }

    void
    operator()( double, state_t const& y, state_t& dy ) const
    {
        double const y1 = y[0];
        double const y2 = y[1];
        double const y3 = y[2];
        double const y4 = y[3];

        double const r1 = sqrt( ( y1 + mu ) * ( y1 + mu ) + y2 * y2 );
        double const r2 = sqrt( ( y1 - 1 + mu ) * ( y1 - 1 + mu ) + y2 * y2 );

        dy[0] = y3;
        dy[1] = y4;
        dy[2] = y1 + 2 * y4 - ( 1 - mu ) * ( y1 + mu ) / ( r1 * r1 * r1 ) - mu * ( y1 - 1 + mu ) / ( r2 * r2 * r2 );
        dy[3] = y2 - 2 * y3 - ( 1 - mu ) * y2 / ( r1 * r1 * r1 ) - mu * y2 / ( r2 * r2 * r2 );
    }
};

int
main( int, char** )
{
    std::string const dirname = "arenstorf_data";
    std::string filename;

    using state_t = std::valarray<double>;

    double const tf = 17.0652165601579625588917206249;
    double const dt = 1e-5;

    double const mu = 0.012277471;

    auto arenstorf_pb = arenstorf_model( mu );

    state_t const yini = { 0.994, 0., 0., -2.00158510637908252240537862224 };

    filename = ( std::filesystem::path( dirname ) / "arenstorf_rk118.dat" ).string();
    ponio::solve( arenstorf_pb, ponio::runge_kutta::rk_118(), yini, { 0., tf }, dt, ponio::observer::file_observer( filename ) );

    filename = ( std::filesystem::path( dirname ) / "arenstorf_rk546m.dat" ).string();
    ponio::solve( arenstorf_pb, ponio::runge_kutta::rk54_6m( 1e-5 ), yini, { 0., tf }, dt, ponio::observer::file_observer( filename ) );

    filename = ( std::filesystem::path( dirname ) / "arenstorf_rk547m.dat" ).string();
    ponio::solve( arenstorf_pb, ponio::runge_kutta::rk54_7m( 1e-5 ), yini, { 0., tf }, dt, ponio::observer::file_observer( filename ) );

    filename = ( std::filesystem::path( dirname ) / "arenstorf_rk547s.dat" ).string();
    ponio::solve( arenstorf_pb, ponio::runge_kutta::rk54_7s( 1e-5 ), yini, { 0., tf }, dt, ponio::observer::file_observer( filename ) );

    // adaptive Strang method with Lipschitz estimate
    using namespace std::placeholders;
    auto phi_1 = [&]( double t, state_t const& y, state_t& dy )
    {
        arenstorf_pb.phi_1( t, y, dy );
    };
    auto phi_2 = [&]( double t, state_t const& y, state_t& dy )
    {
        arenstorf_pb.phi_2( t, y, dy );
    };

    auto arenstorf_splitting_pb = ponio::make_problem( phi_1, phi_2 );

    auto adaptive_strang = ponio::splitting::make_adaptive_strang_tuple( 0.05,
        1e-5,
        std::make_pair( ponio::runge_kutta::rk_44(), 0.5 * dt ),
        std::make_pair( ponio::runge_kutta::rk_44(), 0.5 * dt ) );

    filename       = ( std::filesystem::path( dirname ) / "arenstorf_adaptive_strang.dat" ).string();
    auto sol_range = ponio::make_solver_range( arenstorf_splitting_pb, adaptive_strang, yini, { 0., tf }, dt );
    auto it_sol    = sol_range.begin();

    auto obs = ponio::observer::file_observer( filename );

    static constexpr std::size_t N_delta = 50;
    std::size_t n_iteration              = 0;

    // constant for safety factor
    static constexpr double beta  = 0.1;
    static constexpr double gamma = 0.95;

    double dt_star = 10;
    double C_delta = 0.;
    double C_0     = 0.;

    while ( it_sol->time < tf )
    {
        obs( it_sol->time, it_sol->state, it_sol->time_step );

        ++it_sol;

        if ( n_iteration % N_delta == 0 || ( it_sol->time_step < beta * dt_star || gamma * dt_star < it_sol->time_step ) )
        {
            auto [omega,
                C0] = it_sol.meth.lipschitz_constant_estimate( arenstorf_splitting_pb, it_sol->time, it_sol->state, it_sol->time_step );
            C_0     = C0;

            dt_star = it_sol.meth.info().error / ( C0 * it_sol->time_step * it_sol->time_step );
            C_delta = it_sol.meth.info().error / ( it_sol->time_step * it_sol->time_step * it_sol.meth.info().delta );
        }

        if ( it_sol->time_step < beta * dt_star || gamma * dt_star < it_sol->time_step )
        {
            dt_star                  = it_sol->time_step;
            it_sol.meth.info().delta = dt_star * C_0 / C_delta;
        }

        if ( it_sol.meth.info().success )
        {
            ++n_iteration;
        }
    }

    return 0;
}

File: ponio/examples/arenstorf.py
# Copyright 2022 PONIO TEAM. All rights reserved.
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file.

#!/usr/bin/env python

import subprocess
import os
import dataclasses

import matplotlib.pyplot as plt
import numpy as np
import numpy.typing as npt

name = "arenstorf"
data_dir = name+"_data"


@dataclasses.dataclass
class simu_data:
    label: str
    data: npt.NDArray[np.float64]

    def time(self):
        return self.data[:, 0]

    def pos_x(self):
        return self.data[:, 1]

    def pos_y(self):
        return self.data[:, 2]

    def vel_x(self):
        return self.data[:, 3]

    def vel_y(self):
        return self.data[:, 4]

    def time_step(self):
        return self.data[:, -1]

    def norm_l1(self):
        return np.abs(self.data[:, 1]) + np.abs(self.data[:, 2]) + np.abs(self.data[:, 3]) + np.abs(self.data[:, 4])

    def norm_l2(self):
        return np.sqrt(np.square(self.data[:, 1]) + np.square(self.data[:, 2]) + np.square(self.data[:, 3]) + np.square(self.data[:, 4]))


def load_data(tagname, label):
    return simu_data(label=label, data=np.loadtxt(os.path.join(data_dir, f"{name}_{tagname}.dat")))


def plot_orbit(data, alpha=1):
    plt.plot(data.pos_x(), data.pos_y(), "-+", alpha=alpha, label=data.label)


def plot_velocity(data, alpha=1):
    plt.plot(data.vel_x(), data.vel_y(), "-+", alpha=alpha, label=data.label)


def plot_time_step(data, alpha=1):
    plt.plot(data.time(), data.time_step(),
             "-+", alpha=alpha, label=data.label)


def plot_error(data, ref, alpha=1):
    plt.plot(data.time(), np.abs(data.norm_l2() - np.interp(data.time(),
             ref.time(), ref.norm_l2())), "-+", alpha=alpha, label=data.label)


os.makedirs(data_dir, exist_ok=True)

make = subprocess.Popen(["make", name])
make.wait()

args = [os.path.join(".", name)]
process = subprocess.Popen(args)
process.wait()

rk118 = load_data("rk118", "RK(11, 8)")
rk546m = load_data("rk546m", "RK5(4) 6M")
rk547m = load_data("rk547m", "RK5(4) 7M")
rk547s = load_data("rk547s", "RK5(4) 7S")
strang = load_data("adaptive_strang", "Adaptive Strang")

# orbit plot
plt.rcParams["figure.figsize"] = (6, 6)
plt.axes().set_aspect('equal', 'box')

plot_orbit(rk546m)
plot_orbit(rk547m)
plot_orbit(rk547s)
plot_orbit(strang, 0.05)

plt.axhline(0., 0, 1, linestyle="--", color="grey", linewidth=1)
plt.axvline(0., 0, 1, linestyle="--", color="grey", linewidth=1)

plt.xlim(-1.5, 1.5)
plt.ylim(-1.5, 1.5)

plt.xlabel("$x$")
plt.ylabel("$y$")

plt.tight_layout(pad=1.0)
plt.legend()

plt.savefig(os.path.join(data_dir, "01.png"), dpi=200)

plt.title("Arenstorf orbit")
plt.show()

plt.rcParams["figure.figsize"] = plt.rcParamsDefault["figure.figsize"]

# velocity plot
plt.rcParams["figure.figsize"] = (6, 6)
plt.axes().set_aspect('equal', 'box')

plot_velocity(rk546m)
plot_velocity(rk547m)
plot_velocity(rk547s)
plot_velocity(strang, 0.05)

plt.axhline(0., 0, 1, linestyle="--", color="grey", linewidth=1)
plt.axvline(0., 0, 1, linestyle="--", color="grey", linewidth=1)

plt.xlim(-2.5, 2.5)
plt.ylim(-3.5, 1.5)

plt.xlabel(r"$\dot{{x}}$")
plt.ylabel(r"$\dot{{y}}$")

plt.tight_layout(pad=1.0)
plt.legend()

plt.savefig(os.path.join(data_dir, "02.png"), dpi=200)

plt.title("Arenstorf velocity")
plt.show()

plt.rcParams["figure.figsize"] = plt.rcParamsDefault["figure.figsize"]

# time step history
plot_time_step(rk546m)
plot_time_step(rk547m)
plot_time_step(rk547s)
plot_time_step(strang)

plt.yscale('log')
plt.xlabel(r"$t$")
plt.ylabel(r"$\Delta t$")
plt.legend()

plt.savefig(os.path.join(data_dir, "03.png"))

plt.title("Arenstorf orbit, time step size")
plt.show()

# error throw time
plot_error(rk546m, rk118)
plot_error(rk547m, rk118)
plot_error(rk547s, rk118)
plot_error(strang, rk118, 0.1)

plt.yscale('log')
plt.xlabel(r"$t$")
plt.ylabel(r"$\|\cdot\|_2$")
plt.legend()

plt.savefig(os.path.join(data_dir, "04.png"))

plt.title("Arenstorf orbit, error over time")
plt.show()
